File: pkg/crypto/signer.go
package crypto

import (
	"crypto/ecdsa"
	"crypto/rand"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

// Signer manages ECDSA key pairs for signing transactions
// Uses secp256k1 curve (Ethereum-compatible)
type Signer struct {
	privateKey *ecdsa.PrivateKey
	publicKey  *ecdsa.PublicKey
	address    common.Address
}

// GenerateKey creates a new random secp256k1 key pair
// Returns a Signer with private key, public key, and derived Ethereum address
func GenerateKey() (*Signer, error) {
	privateKey, err := crypto.GenerateKey()
	if err != nil {
		return nil, fmt.Errorf("failed to generate key: %w", err)
	}

	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("failed to cast public key to ECDSA")
	}

	address := crypto.PubkeyToAddress(*publicKeyECDSA)

	return &Signer{
		privateKey: privateKey,
		publicKey:  publicKeyECDSA,
		address:    address,
	}, nil
}

// FromPrivateKeyHex creates a Signer from a hex-encoded private key
// Format: "0x1234..." or "1234..." (64 hex chars)
func FromPrivateKeyHex(hexKey string) (*Signer, error) {
	privateKey, err := crypto.HexToECDSA(hexKey)
	if err != nil {
		return nil, fmt.Errorf("failed to parse private key: %w", err)
	}

	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("failed to cast public key to ECDSA")
	}

	address := crypto.PubkeyToAddress(*publicKeyECDSA)

	return &Signer{
		privateKey: privateKey,
		publicKey:  publicKeyECDSA,
		address:    address,
	}, nil
}

// Address returns the Ethereum address derived from the public key
func (s *Signer) Address() common.Address {
	return s.address
}

// PrivateKeyHex returns the private key as hex string (WITHOUT 0x prefix)
// WARNING: Keep this secret! Never expose to users or logs
func (s *Signer) PrivateKeyHex() string {
	return fmt.Sprintf("%x", crypto.FromECDSA(s.privateKey))
}

// PublicKeyHex returns the public key as hex string (uncompressed, 128 chars)
func (s *Signer) PublicKeyHex() string {
	return fmt.Sprintf("%x", crypto.FromECDSAPub(s.publicKey))
}

// Sign signs a message hash using ECDSA and returns the signature
// Returns signature in [R || S || V] format (65 bytes)
// V is recovery ID (0 or 1) + 27 for Ethereum compatibility
func (s *Signer) Sign(hash []byte) ([]byte, error) {
	if len(hash) != 32 {
		return nil, fmt.Errorf("hash must be 32 bytes, got %d", len(hash))
	}

	signature, err := crypto.Sign(hash, s.privateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to sign: %w", err)
	}

	return signature, nil
}

// SignMessage signs a message (not a hash) by first hashing it with Keccak256
// Use this for arbitrary byte messages
func (s *Signer) SignMessage(message []byte) ([]byte, error) {
	hash := crypto.Keccak256Hash(message)
	return s.Sign(hash.Bytes())
}

// VerifySignature verifies that signature was created by address for given hash
// Returns true if signature is valid, false otherwise
func VerifySignature(address common.Address, hash []byte, signature []byte) bool {
	if len(signature) != 65 {
		return false
	}
	if len(hash) != 32 {
		return false
	}

	// Recover public key from signature
	publicKeyBytes, err := crypto.Ecrecover(hash, signature)
	if err != nil {
		return false
	}

	// Convert to ECDSA public key
	publicKey, err := crypto.UnmarshalPubkey(publicKeyBytes)
	if err != nil {
		return false
	}

	// Derive address from recovered public key
	recoveredAddr := crypto.PubkeyToAddress(*publicKey)

	// Compare addresses
	return recoveredAddr == address
}

// RecoverAddress recovers the signer's address from a message hash and signature
// Returns the address that created the signature
func RecoverAddress(hash []byte, signature []byte) (common.Address, error) {
	if len(signature) != 65 {
		return common.Address{}, fmt.Errorf("invalid signature length: %d", len(signature))
	}
	if len(hash) != 32 {
		return common.Address{}, fmt.Errorf("invalid hash length: %d", len(hash))
	}

	// Recover public key
	publicKeyBytes, err := crypto.Ecrecover(hash, signature)
	if err != nil {
		return common.Address{}, fmt.Errorf("failed to recover public key: %w", err)
	}

	// Convert to ECDSA public key
	publicKey, err := crypto.UnmarshalPubkey(publicKeyBytes)
	if err != nil {
		return common.Address{}, fmt.Errorf("failed to unmarshal public key: %w", err)
	}

	// Derive address
	address := crypto.PubkeyToAddress(*publicKey)
	return address, nil
}

// SignatureToRSV splits a 65-byte signature into R, S, V components
// Useful for Solidity verification or debugging
func SignatureToRSV(signature []byte) (r, s *big.Int, v uint8, err error) {
	if len(signature) != 65 {
		return nil, nil, 0, fmt.Errorf("invalid signature length: %d", len(signature))
	}

	r = new(big.Int).SetBytes(signature[:32])
	s = new(big.Int).SetBytes(signature[32:64])
	v = signature[64]

	return r, s, v, nil
}

// RSVToSignature combines R, S, V into a 65-byte signature
func RSVToSignature(r, s *big.Int, v uint8) []byte {
	signature := make([]byte, 65)
	copy(signature[:32], r.Bytes())
	copy(signature[32:64], s.Bytes())
	signature[64] = v
	return signature
}

// GenerateNonce generates a cryptographically secure random nonce
// Used for replay protection (incremental nonces are preferred, but random works for testing)
func GenerateNonce() (uint64, error) {
	var nonce uint64
	nonceBytes := make([]byte, 8)
	if _, err := rand.Read(nonceBytes); err != nil {
		return 0, fmt.Errorf("failed to generate nonce: %w", err)
	}
	nonce = uint64(nonceBytes[0]) | uint64(nonceBytes[1])<<8 | uint64(nonceBytes[2])<<16 |
		uint64(nonceBytes[3])<<24 | uint64(nonceBytes[4])<<32 | uint64(nonceBytes[5])<<40 |
		uint64(nonceBytes[6])<<48 | uint64(nonceBytes[7])<<56
	return nonce, nil
}

File: src/crypto/signer.rs
//! ECDSA Signing Utilities
//!
//! Provides key generation, signing, and address recovery.

use alloy_primitives::{keccak256, Address, B256};
use k256::ecdsa::{RecoveryId, Signature, SigningKey, VerifyingKey};
use k256::elliptic_curve::rand_core::OsRng;

/// ECDSA signer with private key
pub struct Signer {
    signing_key: SigningKey,
}

impl Signer {
    /// Generate a new random key
    pub fn generate() -> Self {
        let signing_key = SigningKey::random(&mut OsRng);
        Self { signing_key }
    }

    /// Create signer from private key bytes
    pub fn from_bytes(bytes: &[u8; 32]) -> Result<Self, SignerError> {
        let signing_key = SigningKey::from_bytes(bytes.into())
            .map_err(|_| SignerError::InvalidKey)?;
        Ok(Self { signing_key })
    }

    /// Create signer from hex string (with or without 0x prefix)
    pub fn from_hex(hex_str: &str) -> Result<Self, SignerError> {
        let hex_str = hex_str.strip_prefix("0x").unwrap_or(hex_str);
        let bytes = hex::decode(hex_str).map_err(|_| SignerError::InvalidHex)?;
        if bytes.len() != 32 {
            return Err(SignerError::InvalidKey);
        }
        let mut key_bytes = [0u8; 32];
        key_bytes.copy_from_slice(&bytes);
        Self::from_bytes(&key_bytes)
    }

    /// Get the Ethereum address for this signer
    pub fn address(&self) -> Address {
        let verifying_key = self.signing_key.verifying_key();
        public_key_to_address(verifying_key)
    }

    /// Get the private key as bytes
    pub fn private_key(&self) -> [u8; 32] {
        self.signing_key.to_bytes().into()
    }

    /// Get the private key as hex string
    pub fn private_key_hex(&self) -> String {
        format!("0x{}", hex::encode(self.private_key()))
    }

    /// Sign a 32-byte message hash
    /// Returns 65-byte signature (r || s || v)
    pub fn sign(&self, message_hash: &[u8; 32]) -> [u8; 65] {
        let (signature, recovery_id) = self.signing_key
            .sign_prehash_recoverable(message_hash)
            .expect("signing should not fail");

        let mut sig_bytes = [0u8; 65];
        sig_bytes[..64].copy_from_slice(&signature.to_bytes());
        sig_bytes[64] = recovery_id.to_byte() + 27; // Ethereum uses 27/28

        sig_bytes
    }

    /// Sign a message with Ethereum prefix
    /// Adds "\x19Ethereum Signed Message:\n" prefix
    pub fn sign_message(&self, message: &[u8]) -> [u8; 65] {
        let hash = hash_message(message);
        self.sign(&hash)
    }
}

/// Recover address from signature and message hash
pub fn recover_address(message_hash: &[u8; 32], signature: &[u8]) -> Result<Address, SignerError> {
    if signature.len() != 65 {
        return Err(SignerError::InvalidSignature);
    }

    let sig = Signature::try_from(&signature[..64])
        .map_err(|_| SignerError::InvalidSignature)?;

    let v = signature[64];
    let recovery_id = if v >= 27 {
        RecoveryId::try_from(v - 27).map_err(|_| SignerError::InvalidSignature)?
    } else {
        RecoveryId::try_from(v).map_err(|_| SignerError::InvalidSignature)?
    };

    let verifying_key = VerifyingKey::recover_from_prehash(message_hash, &sig, recovery_id)
        .map_err(|_| SignerError::RecoveryFailed)?;

    Ok(public_key_to_address(&verifying_key))
}

/// Hash a message with Ethereum prefix
pub fn hash_message(message: &[u8]) -> [u8; 32] {
    let prefix = format!("\x19Ethereum Signed Message:\n{}", message.len());
    let mut data = prefix.into_bytes();
    data.extend_from_slice(message);
    keccak256(&data).into()
}

/// Convert public key to Ethereum address
fn public_key_to_address(verifying_key: &VerifyingKey) -> Address {
    let public_key = verifying_key.to_encoded_point(false);
    let public_key_bytes = &public_key.as_bytes()[1..]; // Skip 0x04 prefix
    let hash = keccak256(public_key_bytes);
    Address::from_slice(&hash[12..])
}

/// Signer errors
#[derive(Debug, Clone, thiserror::Error)]
pub enum SignerError {
    #[error("invalid private key")]
    InvalidKey,
    #[error("invalid hex string")]
    InvalidHex,
    #[error("invalid signature")]
    InvalidSignature,
    #[error("recovery failed")]
    RecoveryFailed,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_and_sign() {
        let signer = Signer::generate();
        let message = b"hello world";
        let hash = keccak256(message);
        let signature = signer.sign(&hash.into());

        assert_eq!(signature.len(), 65);

        // Recover address
        let recovered = recover_address(&hash.into(), &signature).unwrap();
        assert_eq!(recovered, signer.address());
    }

    #[test]
    fn test_from_hex() {
        let key_hex = "0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
        let signer = Signer::from_hex(key_hex).unwrap();
        assert!(!signer.address().is_zero());
    }

    #[test]
    fn test_address_derivation() {
        // Known test vector
        let key_hex = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
        let signer = Signer::from_hex(key_hex).unwrap();
        // This is the first Anvil/Hardhat account
        assert_eq!(
            signer.address().to_string().to_lowercase(),
            "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"
        );
    }
}
